### ISSUE TITLE
[BUG] CDS/LDS names must not exceed 60 chars

### ISSUE THREAD

**Issue by prune998** [istio/istio#2735]
Istio Version : 0.4.0 (mixer, pilot, proxy)

When upgrading from 0.2.x to 0.4.0, I have this issue when the proxy starts : 
```
ctxerd-04428f0-86bbd689b6-5wcm9 istio-proxy 2018-01-22T15:39:31.826727805Z [2018-01-22 15:39:31.826][16][warning][upstream] external/envoy/source/common/upstream/cds_subscription.cc:65] cds: fetch failure: JSON at lines 128-134 does not conform to schema.
ctxerd-04428f0-86bbd689b6-5wcm9 istio-proxy 2018-01-22T15:39:31.826776462Z  Invalid schema: #/properties/name
ctxerd-04428f0-86bbd689b6-5wcm9 istio-proxy 2018-01-22T15:39:31.826782306Z  Schema violation: maxLength
ctxerd-04428f0-86bbd689b6-5wcm9 istio-proxy 2018-01-22T15:39:31.82678674Z  Offending document key: #/name
```

I tried to check CDS using curl : 
```
curl  http://istio-pilot.istio-system:8080/v1/clusters/istio-proxy/sidecar~10.20.5.9~ctxerd-04428f0-86bbd689b6-5wcm9.staging~staging.svc.cluster.local
```

Line 128 seems to be in the block : 
```
   {
    "name": "out.elasticsearch-elasticsearch-head.monitoring.svc.cluster.local|http",
    "service_name": "elasticsearch-elasticsearch-head.monitoring.svc.cluster.local|http",
    "connect_timeout_ms": 1000,
    "type": "sds",
    "lb_type": "round_robin"
   },
```

As far as I remember the RFC for a FQDN is 255 bytes... Seems this limitation wasn't there in 0.2.x

**Comment by prune998** [comment]
0.2.19 Pilot was responding : 
```
  {    
    "name": "out.a36d97a125dc0383e2682825450b2ab29e81f753",
    "service_name": "elasticsearch-elasticsearch-head.monitoring.svc.cluster.local|http",
    "connect_timeout_ms": 1000,
    "type": "sds",
    "lb_type": "round_robin"
   },
```

Which was OK for Envoy...

**Comment by prune998** [comment]
Sounds it's linked to this Envoy code : 
https://github.com/envoyproxy/envoy/pull/667/files/1af0ddad80ae7f41c965dd378070191ea55ff36c

It's clearly stated that `The cluster name can be at most 60 characters long, and must not contain ``:``.`

**Comment by rshriram** [comment]
This is very odd. This issue was fixed. Our cluster names are 181 characters in length. I fixed it and added the option in pilot-agent to start envoy with the special flag (--max-cluster-name-length thing). Can you dump the envoy CLI args?

**Comment by prune998** [comment]
`pilot-agent` is started with  : 
```
Commands:
-
Args:
proxy
sidecar
-v
2
--configPath
/etc/istio/proxy
--binaryPath
/usr/local/bin/envoy
--serviceCluster
useredged
--drainDuration
45s
--parentShutdownDuration
1m0s
--discoveryAddress
istio-pilot.istio-system:8080
--discoveryRefreshDelay
1s
--zipkinAddress
jaeger-collector.monitoring:9411
--connectTimeout
10s
--statsdUdpAddress
istio-mixer.istio-system:9125
--proxyAdminPort
15000
--controlPlaneAuthPolicy
NONE
```
`envoy` is started with : 
```
/usr/local/bin/envoy -c /etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster useredged --service-node sidecar~10.20.1.25~useredged-d11cfdc-6b8d96f879-6mhg8.staging~staging.svc.cluster.local --max-obj-name-len 189 --service-zone AvailabilityZone couldn't find the given cluster node
```

So the `max-obj-name-len 189` is specified. Strange startup for `--service-zone`, will have to investigate. 

istio proxy is version 0.4.0, injected using `istioctl` 0.4.0. Kubernetes 1.8.6 on GKE

I will re-create the offending application with a long name and see what happen. The logs may have appeared during my upgrade from 0.2.19 to 0.4.0... 
Thanks for your answer @rshriram

**Comment by ijsnellf** [comment]
Should be fixed in 0.5.1: https://github.com/istio/istio/issues/2792
